source/common/router/router.cc

```diff
@@ -1019,6 +1035,14 @@ void Filter::UpstreamRequest::encodeData(Buffer::Instance& data, bool end_stream
 
     buffered_request_body_->move(data);
   } else {
+    // Encodes metadata if exists.
+    if (!parent_.downstream_metadata_map_vector_.empty()) {
+      ENVOY_STREAM_LOG(trace, "Send metadata before sending data. {}", *parent_.callbacks_,
+                       parent_.downstream_metadata_map_vector_);
+      request_encoder_->encodeMetadata(parent_.downstream_metadata_map_vector_);
+      parent_.downstream_metadata_map_vector_.clear();
```

Comment:
Great point!! Thanks! I didn't think of those cases at all.

I am still working on this part. But I agree we should have a design for them before submitting this PR in case there are changes. Will update this thread when I make progress.

Reply:
I only have looked at the internal redirect case. Since those cases all call recreateStream(), I think they can be solved similarly.

recreateStream() calls decodeHeaders() to go through all filters at the end of the the function. https://github.com/envoyproxy/envoy/blob/4f4280ccd00b0e3b057f4c82b13bf6c4513c4ef2/source/common/http/conn_manager_impl.cc#L1789 . This allows us to add metadata again for the redirected request, so I think we may not need extra handling for the redirect case in envoy.

The issue is in nghttp2. In the redirect case, metadata ends up in session->ob_reg: https://github.com/nghttp2/nghttp2/blob/d93842db3eb481fc6953ec70a81ed2f7863ffb45/lib/nghttp2_session.c#L2352, and the redirect headers ends up in session->ob_syn: https://github.com/nghttp2/nghttp2/blob/d93842db3eb481fc6953ec70a81ed2f7863ffb45/lib/nghttp2_session.c#L2360 . Because session->ob_reg is parsed before session->ob_syn, metadata is received before headers (no new stream has created), and caused failure.

I switched the positions of the pointed code, and I don't see new test failures in nghttp2. And a simple internal redirect test case passed with metadata in envoy. (cl/231599606. If others are interested in the change, I can make a public available PR).

What do you think about making the nghttp2 change? If we go with this, we can submit this PR first, and deal with shadowing, internal rediects and retries after the change is pushed to nghttp2. But it may take longer. I notice envoy still uses an old version of nghttp2.

Reply:
To add more details about why the nghttp2 fix ([here](https://github.com/nghttp2/nghttp2/pull/1312)) is needed in redirect, etc cases, but not needed in regular case. 

In the regular case, when we call encodeHeaders() in onPoolReady(), we will submit the headers by calling sendPendingFrames() before encoding any other frames. Calling sendPendingFrames() will make sure headers (session->ob_syn) are popped for processing first, because that's the only frame available. In redirect case, sendPendingFrames() in encodeHeaders() will be skipped because we are still dispatching (dispatching_ is true [here](https://github.com/envoyproxy/envoy/blob/dc3467adacb8f2b0393583857e709c06d76a2441/source/common/http/http2/codec_impl.cc#L671)). So we accumulate both headers and metadata. When sendPendingFrames is executed at the end of dispatch(), metadata shows up first because it is in session->ob_reg ([here](https://github.com/nghttp2/nghttp2/blob/d93842db3eb481fc6953ec70a81ed2f7863ffb45/lib/nghttp2_session.c#L2352)), which is processed before session->ob_syn([here](https://github.com/nghttp2/nghttp2/blob/d93842db3eb481fc6953ec70a81ed2f7863ffb45/lib/nghttp2_session.c#L2360)).

It's easiest, and I think it also makes sense to fix in nghttp2. (sync frame should always process before regular frames?). If it takes too long, I will try to find some hacky ways to fix it in Envoy.

Reply:
As said I'm totally fine dealing with this later, I'd just like a tracking TODO :-)

Reply:
Fixed. Thanks!